README.md
## AGBI Hackathon - Team Appendly 🚀
![forthebadge](https://forthebadge.com/images/badges/built-with-love.svg)
![forthebadge](https://forthebadge.com/images/badges/made-with-javascript.svg)
![forthebadge](https://forthebadge.com/images/badges/made-with-python.svg)

<img src="https://media-fastly.hackerearth.com/media/hackathon/agbi-digital-healthtech-grand-challenge/images/4e1412d2d7-great_lakes_he_cover-01-01-01-01-01-01-01-01.png" width="100%">

### Problem Statement 🤔:
* ### Diagnosis Tagging for Digital Marketing:
* Currently, hospitals rely on mass marketing campaigns to inform patients of a relevant treatment or drug available to further improve their health. Build applications that can help hospitals, patients, pharmaceutical companies, and health R&D companies with more efficient and effective ways of communication by hospitals for better health outcomes and targeted communication to prime patient groups for a new treatment/drug leveraging the power of data analytics. Your solution needs to manage the database of patient records and discharge summaries to identify patients who are good candidates for newly launched medicines, treatments etc.

### Objectives ✔️:
* To convert decades of handwritten hospital documents and reports into text and store it in a database for analytics.
* To visualize the data and derive insights out of them for better health outcomes.
* To have the database of patients and drugs at one place.
* To apply filter on those data and implement targeted marketing of new drugs and treatments.
* To establish a perennial digital bridge between patients and hospital

### Our Solution 👨‍💻:
<img src="https://github.com/Sai-Adarsh/agbi-hackathon/blob/main/diagnosis-tagging/app/base/static/assets/img/icon.png" width="30%">

* Mehta's Digital Marketing [Webapp](https://github.com/Sai-Adarsh/agbi-hackathon/tree/main/diagnosis-tagging)
* Mehta's OCRApp
* Mehta's PatientsApp

### Demo URL 📱:
* Mehta's Digital Marketing: [webapp](https://agbihackathon.herokuapp.com/) 💻
* Mehta's OCRApp [apk](https://github.com/Sai-Adarsh/agbi-hackathon/raw/main/ocrapp/build/mehtaocrapp.apk) • [snack](https://expo.io/@saiadarsh99/mehta-ocr) 📱
* Mehta's PatientsApp [apk](https://github.com/Sai-Adarsh/agbi-hackathon/raw/main/patients-sms-app/android/app/patients-sms-app.apk) 📱

### Architecture 😎: 

<img src="https://github.com/Sai-Adarsh/agbi-hackathon/blob/main/diagnosis-tagging/architecture.png" width="90%">
<img src="https://github.com/Sai-Adarsh/agbi-hackathon/blob/main/diagnosis-tagging/architecture(1).png" width="90%">

### Tech Stacks ⚛️:

<img src="https://github.com/Sai-Adarsh/agbi-hackathon/blob/main/diagnosis-tagging/stack.png" width="90%">

### Installation 📦:
* ## Digital Marketing Webapp
* Clone the repo
```sh
   $ git clone https://github.com/Sai-Adarsh/agbi-hackathon
```
```sh
   $ cd agbi-hackathon/diagnosis-tagging
```
* Create virtualenv
```sh
   $ virtualenv venv
```
* For Linux
```sh
   $ source ./venv/bin/activate
```
* For windows
```sh
   $ cd venv/Scripts/
```
```sh
   $ activate
```
* Install dependencies
```sh
   $ pip install -r requirements.txt
```
* ### How to run 💡:
* Run the webapp locally
```sh
   $ python run.py
```
* Now, the application runs in [localhost](http://127.0.0.1:5000/)
* ## OCR App
```sh
   $ cd agbi-hackathon/ocrapp
```
```sh
   $ yarn
```
```sh
   $ yarn start
```
* ## Patients App
```sh
   $ cd agbi-hackathon/patients-sms-app
```
```sh
   $ flutter run
```

## Instructions 📋:
### OCR App 📱
* There are two ways to use the OCR App:
* Tap the camera icon -> take a photo of the target -> tap convert
* Tap the file icon -> choose photo from the gallery -> tap convert
* If you wish to add the text/photo to the Firebase:
* Tap push to database to store the image/text in cloud.

### Web App 💻
* Initially register yourself in the Web App.
* Login using your registered credentials.
* Navigate to any of the pages using the Navigation bar in the left.
* Click the buttons 'add drug details', 'add patient details' in the respective tabs to add their details to the firebase.
* Use the text box below the table headers to search/filter the drug/patient details accordingly.
* Use the Send Notification tab to generate a list of patients taking a particular medication by clicking the generate button beside their details.
* Click the button fire notification to send notification to the target group.
* You can also view the details of each patient / drug / your profile by clicking the respective links.
* Logout tab logs you out of the application.

### Patients App 📱
* Just opening the app will fetch you all the messages from the Hospital side.


### Authors 📋:
<p> Made with ❤</p>

* Sai Adarsh
* Rajasekar
* LS Meenatchi

[<img src="https://image.flaticon.com/icons/svg/185/185961.svg" width="35" padding="10">](https://twitter.com/ad6rsh)
[<img src="https://image.flaticon.com/icons/svg/185/185964.svg" width="35" padding="10">](https://www.linkedin.com/in/sai-adarsh/)
[<img src="https://image.flaticon.com/icons/svg/185/185981.svg" width="35" padding="10">](https://www.facebook.com/saiadarsh99)


diagnosis-tagging/CHANGELOG.md
# Change Log

## Unreleased
### Improvements

- 2020-09-11 - Codebase & UI Update
    - Integrate latest UI release & latest Flask Codebase

- 2020-08-20 - Added get_segment() helper that detects the current page
    - Updated files(s): app/home/routes.py

- 2020-08-20 - Added get_segment() helper that detects the current page
    - Updated files(s): app/home/routes.py

- 2020-06-22 - Guard Flask links with quotes
    - Sample href="{{ url_for('base_blueprint.login') }}"
    - Impacted files: login.html, register.html, sidebar.html

- 2020-06-22 - Added HEROKU support. Impacted files:
    - runtime.txt - Bump the Python version to 3.6.10
    - README added new section for HEROKU deployment

- 2020-05-30 - Improvements & Bug Fixes
    - Patch #Bug - Return a 403 Error for unauthorized access
    - Update Licensing information
    - Add CHANGELOG.md to track all changes

## [1.0.0] 2020-02-07
### Initial Release


diagnosis-tagging/mail.py
import smtplib
import ssl
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import csv
import time

sender_email = "saiadarshsivakumar@gmail.com"
# receiver_email = "receiver_mail@gmail.com"
password = "thalathala57"

message = MIMEMultipart("alternative")
message["Subject"] = "Your Mail Subject"
message["From"] = sender_email

context = ssl.create_default_context()
server = smtplib.SMTP_SSL("smtp.gmail.com", 465, context=context)
server.ehlo()
server.login(sender_email, password)

with open('mail.html', 'r', encoding="utf8") as file:
    data = file.read().replace('\n', '')
count = 0

reader = [['Sai Adarsh', 'sivakumar.sk11865.sk@gmail.com', '#'], ['Rajasekar', 'rajasekar1999@gmail.com', '#']]
for x in reader:
    L = ", ".join(x).split(', ')
    name, email, link = L[0], L[1], L[2]
    # Create the plain-text and HTML version of your message
    html = data.format(name=name, link=link)

    # Add HTML/plain-text parts to MIMEMultipart message
    # The email client will try to render the last part first
    message.attach(MIMEText(html, "html"))

    server.sendmail(
        sender_email, email, message.as_string()
    )

    count += 1
    print(str(count) + ". Sent to " + email)

    if(count%80 == 0):
        server.quit()
        print("Server cooldown for 100 seconds")
        time.sleep(100)
        server.ehlo()
        server.login(sender_email, password)

server.quit()

patients-sms-app/lib/main.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';
import 'package:flutter_app/permissions.dart';
import 'package:flutter_app/pit_sms_call_log.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List<SmsLog> messages;
  List<String> messagesToDisplay;
  int length = 0;

  void getSMS() async {
    messages = await PitSmsCallLog.getSmsLog(daysBefore: 10);
  }

  void filter() {
    if (messages != null) {
      messagesToDisplay = new List<String>(messages.length);
      int j = 0;
      for (int i = 0; i < messages.length; i++) {
        if (messages[i].body.contains("9566761938", 0)) {
          messagesToDisplay[j] = messages[i].body;
          j++;
        }
      }
      setState(() {
        length = j;
      });
    } else {
      setState(() {
        getSMS();
      });
    }
  }

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    getSMSPermission();
    setState(() {
      getSMS();
    });
  }

  @override
  Widget build(BuildContext context) {
    filter();
    Future.delayed(Duration(milliseconds: 500), () {
      filter();
      Future.delayed(Duration(milliseconds: 500), () {
        filter();
      });
    });
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text("Notifications"),
          backgroundColor: Colors.black,
        ),
        body: Container(
          color: Colors.black,
          child: ListView.builder(
            itemCount: length,
            itemBuilder: (context, index) {
              return Container(
                margin:
                    EdgeInsets.all(MediaQuery.of(context).size.width * 0.05),
                decoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(15)),
                height: MediaQuery.of(context).size.width * 0.42,
                width: MediaQuery.of(context).size.width * 0.8,
                padding: EdgeInsets.all(15),
                child: Center(
                    child: Text(
                  messagesToDisplay[index],
                )),
              );
            },
          ),
        ),
      ),
    );
  }
}


ocrapp/App.js
import * as ImagePicker from 'expo-image-picker';
import * as Permissions from 'expo-permissions';
import * as firebase from 'firebase';
import { LinearGradient } from 'expo-linear-gradient';
import React, { Component } from 'react';
import {
  ActivityIndicator,
  Button,
  Clipboard,
  Image,
  ToastAndroid,
  ScrollView,
  Share,
  StatusBar,
  TouchableOpacity,
  StyleSheet,
  Text,
  View,
  Dimensions,
} from 'react-native';
import uuid from 'uuid';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { TextInput } from 'react-native-paper';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { SimpleCard } from "@paraboly/react-native-card";

const windowWidth = Dimensions.get('window').width;
const windowHeight = Dimensions.get('window').height;


const Stack = createStackNavigator();
const Tab = createBottomTabNavigator();

var firebaseConfig = {
  apiKey: "AIzaSyD4ibuxkZxlOfCtJvcSHy4HowiTxcbSZKg",
  authDomain: "bubo-3b1fb.firebaseapp.com",
  databaseURL: "https://bubo-3b1fb.firebaseio.com",
  projectId: "bubo-3b1fb",
  storageBucket: "bubo-3b1fb.appspot.com",
  messagingSenderId: "177543367652",
  appId: "1:177543367652:web:7e948c148c61ef75a295f1",
  measurementId: "G-VJQV2D8X5T"
};

if (!firebase.apps.length) {
  firebase.initializeApp(firebaseConfig);
}

// Firebase sets some timeers for a long period, which will trigger some warnings. Let's turn that off for this example
console.disableYellowBox = true;

class App extends Component {
  state = {
    image: null,
    uploading: false,
    googleResponse: null,
  };

  async componentDidMount() {
    await Permissions.askAsync(Permissions.CAMERA_ROLL);
    await Permissions.askAsync(Permissions.CAMERA);
  }

  submitToGoogle = async () => {
      try {
        this.setState({ uploading: true });
        let { image } = this.state;
        let body = JSON.stringify({
          requests: [
            {
              features: [
                { type: "DOCUMENT_TEXT_DETECTION", maxResults: 5 },
              ],
              image: {
                source: {
                  imageUri: image
                }
              }
            }
          ]
        });
        let response = await fetch(
          "https://vision.googleapis.com/v1/images:annotate?key=" +
          "AIzaSyD4ibuxkZxlOfCtJvcSHy4HowiTxcbSZKg",
          {
            headers: {
              Accept: "application/json",
              "Content-Type": "application/json"
            },
            method: "POST",
            body: body
          }
        );
        let responseJson = await response.json();
        responseJson = responseJson.responses[0].fullTextAnnotation.text;
        this.setState({
          googleResponse: responseJson,
          uploading: false
        });
        } catch (error) {
          console.log(error);
        }
    };

  render() {
    let { image } = this.state;

    return (
      <View style={styles.container}>
        <Text>{"\n\n"}</Text>
        <TouchableOpacity
          style={{
           height: 100,
           width: windowWidth - 50,
           borderRadius: 20,
           alignItems: 'center',
           justifyContent: 'center',
           shadowColor: "#000",
           shadowOffset: {
           	width: 0,
           	height: 6,
           },
           shadowOpacity: 0.39,
           shadowRadius: 8.30,

           elevation: 13,
         }}>
         <LinearGradient colors={['#3499FF', '#3A3985']} style={{
           borderRadius: 20,
           height: 100,
           width: windowWidth - 50,
           alignItems: 'center',
           justifyContent: 'center'
         }} start={{ x: 0, y: 1 }} end={{ x: 1, y: 0 }}>
           <Image
             style={{height: 70, width: 250}}
             source = {require('./assets/login.png')}
           />
        </LinearGradient>
        </TouchableOpacity>
        <Text>{"\n"}</Text>
        <TouchableOpacity
          style={{
           height: 200,
           width: windowWidth - 50,
           borderRadius: 20,
           alignItems: 'center',
           justifyContent: 'center',
           shadowColor: "#000",
           shadowOffset: {
           	width: 0,
           	height: 6,
           },
           shadowOpacity: 0.39,
           shadowRadius: 8.30,

           elevation: 13,
         }}>
         <LinearGradient colors={['#121212', '#121212']} style={{
           borderRadius: 20,
           height: 200,
           width: windowWidth - 50,
           alignItems: 'center',
           justifyContent: 'center'
         }} start={{ x: 0, y: 1 }} end={{ x: 1, y: 0 }}>
           <Text style={{fontWeight: 'bold', marginLeft: 20, marginRight: 20, fontSize: 15, color: '#FFF', textAlign: 'center'}}>OCR App by team Appendly - built for AGBI Hackathon. {"\n"}</Text>
           <Text style={{fontWeight: 'bold', marginLeft: 20, marginRight: 20, fontSize: 15 , color: '#FFF', textAlign: 'center'}}>An Application that enables the user to convert any form of handwritten or printed text into editable digital text by either directly taking photo of it or by using a image from the gallery. </Text>
        </LinearGradient>
        </TouchableOpacity>
        <Text>{"\n"}</Text>
        {this._maybeRenderUploadingOverlay()}
        <View style={{
          height: 200,
          width: windowWidth - 50,
          borderRadius: 20,
          alignItems: 'center',
          justifyContent: 'center',
          shadowColor: "#000",
          shadowOffset: {
          	width: 0,
          	height: 6,
          },
          shadowOpacity: 0.39,
          shadowRadius: 8.30,

          elevation: 13,
        }}>
        <LinearGradient colors={['#121212', '#121212']} style={{
          borderRadius: 20,
          height: 200,
          width: windowWidth - 50,
          alignItems: 'center',
          justifyContent: 'center'
        }} start={{ x: 0, y: 1 }} end={{ x: 1, y: 0 }}>
        <TouchableOpacity
        style={styles.button}
        onPress={this._pickImage}>
        <LinearGradient colors={['#3499FF', '#3A3985']} style={styles.button} start={{ x: 0, y: 1 }} end={{ x: 1, y: 1 }}>
          <MaterialCommunityIcons name="file" style={{color: '#fff', alignSelf: 'center'}} size={30} />
        </LinearGradient>
        </TouchableOpacity>
        <Text>{"\n"}</Text>
        <TouchableOpacity
        style={styles.button}
        onPress={this._takePhoto}>
        <LinearGradient colors={['#3499FF', '#3A3985']} style={styles.button} start={{ x: 0, y: 1 }} end={{ x: 1, y: 1 }}>
          <MaterialCommunityIcons name="camera" style={{color: '#fff', alignSelf: 'center'}} size={30} />
        </LinearGradient>

        </TouchableOpacity>
        </LinearGradient>
        </View>

        <StatusBar barStyle="default" />
      </View>
    );
  }

  _maybeRenderUploadingOverlay = () => {
    if (this.state.uploading) {
      return (
        <View
          style={[
            StyleSheet.absoluteFill,
            {
              backgroundColor: 'rgba(0,0,0,0.4)',
              alignItems: 'center',
              justifyContent: 'center',
            },
          ]}>
          <ActivityIndicator color="#000" animating size="large" />
        </View>
      );
    }
  };

  _maybeRenderImage = () => {
    let { image } = this.state;
    if (!image) {
      return;
    }

    return (
      <View
        style={{
          marginTop: 30,
          width: 250,
          borderRadius: 3,
          elevation: 2,
        }}>
        <View
          style={{
            borderTopRightRadius: 3,
            borderTopLeftRadius: 3,
            shadowColor: 'rgba(0,0,0,1)',
            shadowOpacity: 0.2,
            shadowOffset: { width: 4, height: 4 },
            shadowRadius: 5,
            overflow: 'hidden',
          }}>
          <Image source={{ uri: image }} style={{ width: 250, height: 250 }} />
        </View>
        <Text
          onPress={this._copyToClipboard}
          onLongPress={this._share}
          style={{ paddingVertical: 10, paddingHorizontal: 10 }}>
          {image}
        </Text>
      </View>
    );
  };

  _share = () => {
    Share.share({
      message: this.state.image,
      title: 'Check out this photo',
      url: this.state.image,
    });
  };

  _copyToClipboard = () => {
    Clipboard.setString(this.state.image);
    alert('Copied image URL to clipboard');
  };

  _takePhoto = async () => {
    let pickerResult = await ImagePicker.launchCameraAsync({
      allowsEditing: true,
      aspect: [4, 3],
    });

    this._handleImagePicked(pickerResult);
  };

  _pickImage = async () => {
    let pickerResult = await ImagePicker.launchImageLibraryAsync({
      allowsEditing: true,
      aspect: [4, 3],
    });

    this._handleImagePicked(pickerResult);
  };

  _handleImagePicked = async pickerResult => {
    try {
      this.setState({ uploading: true });

      if (!pickerResult.cancelled) {
        uploadUrl = await uploadImageAsync(pickerResult.uri);
        this.setState({ image: uploadUrl });
      }
    } catch (e) {
      console.log(e);
      alert('Upload failed, sorry :(');
    } finally {
      this.setState({ uploading: false });
    }
    await this.submitToGoogle();
    global.myVar = this.state.googleResponse;
    this.setState({ uploading: false });
    this.props.navigation.navigate('FormScreen');
  };
}

global.myVar = null;
class FormScreen extends Component {
  constructor(props) {
    super(props)
    this.state = {
      text: '',
    };
  }

  componentDidMount() {
    this.setState({text: global.myVar});
  }

  pushSampleData () {
    ToastAndroid.show("Updating...", ToastAndroid.SHORT);
    const data = {
      'text': this.state.text,
    };
    const db = firebase.database().ref().child("ocr_records").push(data);
    this.setState({text: ''});
    const navigateToStack = this.props.navigation.navigate('Home');
  }


  render() {
    return (
      <ScrollView
        keyboardShouldPersistTaps='always'
      >
      <View style={styles.postContainer}>
        <Text>{"\n\n"}</Text>
        <Text
          style = {{
            alignSelf: 'center',
            fontSize: 30,
            fontWeight: 'bold',
          }}
        >OCR text{"\n"}</Text>
        <TextInput
          style = {{width: windowWidth - 50, height: windowHeight - 300, textAlignVertical: "top"}}
          label="Text"
          multiline={true}
          type='outlined'
          value={this.state.text}
          onChangeText={text => this.setState({ text })}
        />
        <Text>{"\n"}</Text>
        <TouchableOpacity
          style={styles.button}
          onPress={() => this.pushSampleData()}>
          <LinearGradient colors={['#F54B64', '#F78361']} style={styles.button} start={{ x: 0, y: 1 }} end={{ x: 1, y: 1 }}>
            <Text style={styles.buttonText}> Push to database </Text>
          </LinearGradient>
        </TouchableOpacity>
        <Text>{"\n\n"}</Text>
      </View>
      </ScrollView>
    );
  }
}


class AboutScreen extends Component {
  state = {
    image: null,
    uploading: false,
    googleResponse: null,
  };

  async componentDidMount() {
    await Permissions.askAsync(Permissions.CAMERA_ROLL);
    await Permissions.askAsync(Permissions.CAMERA);
  }

  submitToGoogle = async () => {
      try {
        this.setState({ uploading: true });
        let { image } = this.state;
        let body = JSON.stringify({
          requests: [
            {
              features: [
                { type: "DOCUMENT_TEXT_DETECTION", maxResults: 5 },
              ],
              image: {
                source: {
                  imageUri: image
                }
              }
            }
          ]
        });
        let response = await fetch(
          "https://vision.googleapis.com/v1/images:annotate?key=" +
          "AIzaSyD4ibuxkZxlOfCtJvcSHy4HowiTxcbSZKg",
          {
            headers: {
              Accept: "application/json",
              "Content-Type": "application/json"
            },
            method: "POST",
            body: body
          }
        );
        let responseJson = await response.json();
        responseJson = responseJson.responses[0].fullTextAnnotation.text;
        this.setState({
          googleResponse: responseJson,
          uploading: false
        });
        } catch (error) {
          console.log(error);
        }
    };

  render() {
    let { image } = this.state;

    return (
      <View style={styles.container}>
        <Text>{"\n\n"}</Text>
        <Text style={{color: "#fff", fontSize: 20, fontWeight: 'bold'}}>Start scanning{"\n"}</Text>
        {this._maybeRenderUploadingOverlay()}
        <View style={{
          height: 200,
          width: windowWidth - 50,
          borderRadius: 20,
          alignItems: 'center',
          justifyContent: 'center',
          shadowColor: "#000",
          shadowOffset: {
          	width: 0,
          	height: 6,
          },
          shadowOpacity: 0.39,
          shadowRadius: 8.30,

          elevation: 13,
        }}>
        <LinearGradient colors={['#121212', '#121212']} style={{
          borderRadius: 20,
          height: 200,
          width: windowWidth - 50,
          alignItems: 'center',
          justifyContent: 'center'
        }} start={{ x: 0, y: 1 }} end={{ x: 1, y: 0 }}>
        <TouchableOpacity
        style={styles.button}
        onPress={this._pickImage}>
        <LinearGradient colors={['#3499FF', '#3A3985']} style={styles.button} start={{ x: 0, y: 1 }} end={{ x: 1, y: 1 }}>
          <MaterialCommunityIcons name="file" style={{color: '#fff', alignSelf: 'center'}} size={30} />
        </LinearGradient>
        </TouchableOpacity>
        <Text>{"\n"}</Text>
        <TouchableOpacity
        style={styles.button}
        onPress={this._takePhoto}>
        <LinearGradient colors={['#3499FF', '#3A3985']} style={styles.button} start={{ x: 0, y: 1 }} end={{ x: 1, y: 1 }}>
          <MaterialCommunityIcons name="camera" style={{color: '#fff', alignSelf: 'center'}} size={30} />
        </LinearGradient>

        </TouchableOpacity>
        </LinearGradient>
        </View>

        <StatusBar barStyle="default" />
      </View>
    );
  }

  _maybeRenderUploadingOverlay = () => {
    if (this.state.uploading) {
      return (
        <View
          style={[
            StyleSheet.absoluteFill,
            {
              backgroundColor: 'rgba(0,0,0,0.4)',
              alignItems: 'center',
              justifyContent: 'center',
            },
          ]}>
          <ActivityIndicator color="#000" animating size="large" />
        </View>
      );
    }
  };

  _maybeRenderImage = () => {
    let { image } = this.state;
    if (!image) {
      return;
    }

    return (
      <View
        style={{
          marginTop: 30,
          width: 250,
          borderRadius: 3,
          elevation: 2,
        }}>
        <View
          style={{
            borderTopRightRadius: 3,
            borderTopLeftRadius: 3,
            shadowColor: 'rgba(0,0,0,1)',
            shadowOpacity: 0.2,
            shadowOffset: { width: 4, height: 4 },
            shadowRadius: 5,
            overflow: 'hidden',
          }}>
          <Image source={{ uri: image }} style={{ width: 250, height: 250 }} />
        </View>
        <Text
          onPress={this._copyToClipboard}
          onLongPress={this._share}
          style={{ paddingVertical: 10, paddingHorizontal: 10 }}>
          {image}
        </Text>
      </View>
    );
  };

  _share = () => {
    Share.share({
      message: this.state.image,
      title: 'Check out this photo',
      url: this.state.image,
    });
  };

  _copyToClipboard = () => {
    Clipboard.setString(this.state.image);
    alert('Copied image URL to clipboard');
  };

  _takePhoto = async () => {
    let pickerResult = await ImagePicker.launchCameraAsync({
      allowsEditing: true,
      aspect: [4, 3],
    });

    this._handleImagePicked(pickerResult);
  };

  _pickImage = async () => {
    let pickerResult = await ImagePicker.launchImageLibraryAsync({
      allowsEditing: true,
      aspect: [4, 3],
    });

    this._handleImagePicked(pickerResult);
  };

  _handleImagePicked = async pickerResult => {
    try {
      this.setState({ uploading: true });

      if (!pickerResult.cancelled) {
        uploadUrl = await uploadImageAsync(pickerResult.uri);
        this.setState({ image: uploadUrl });
      }
    } catch (e) {
      console.log(e);
      alert('Upload failed, sorry :(');
    } finally {
      this.setState({ uploading: false });
    }
    await this.submitToGoogle();
    global.myVar = this.state.googleResponse;
    this.setState({ uploading: false });
    this.props.navigation.navigate('FormScreen');
  };
}


class StackScreen extends Component {
  render () {
    return (
        <Stack.Navigator
          screenOptions={{
            headerShown: false,
            headerStyle: {
              backgroundColor: '#002bff',
              shadowColor: 'transparent',
              elevation: 0,
            },
            headerTintColor: '#FFF',
            headerTitleAlign: 'center',
            headerTitleStyle: {
              //paddingBottom: 10
            },
          }}>
          <Stack.Screen
            name="Home" component={App}/>
            <Stack.Screen
            name="FormScreen" component={FormScreen}/>
        </Stack.Navigator>
    );
  }
}


class DeveloperScreen extends Component {
  render () {
    return (
      <View style={styles.container}>
      <TouchableOpacity>
        <Image
             style={{
               borderRadius: 200,
               width: 150,
               height: 150,
             }}
             source={require('./assets/developer.jpg')}
         />
      </TouchableOpacity>
      <Text>{"\n"}</Text>
      <View style={{ marginTop: 16 }}>
        <SimpleCard backgroundColor="#212121" titleTextColor="#FFF" title="This app was developed by Team Appendly for AGBI Hackathon 2020. Source code copyrights © 2020 Appendly."/>
      </View>
      </View>
    );
  };
}

class AppScreen extends Component {
  render () {
    return (
      <NavigationContainer>
        <Tab.Navigator
          initialRouteName="Stack"
          tabBarOptions={{
            style: {
              borderTopColor: '#transparent',
            },
            labelStyle: {
              fontSize: 10,
            },
            activeBackgroundColor: '#1e1e1e',
            inactiveBackgroundColor: '#1e1e1e',
            inactiveTintColor: '#babbc0',
            activeTintColor: '#3499FF',
          }}
        >
          <Tab.Screen name="Stack" component={StackScreen}
            options= {{
              tabBarLabel: 'Home',
              tabBarIcon: ({ color, size }) => (
                  <MaterialCommunityIcons name="home" color={color} size={size} />
              ),
            }}
          />
          <Tab.Screen name="Post" component={AboutScreen}
            options= {{
              tabBarLabel: 'Scan',
              tabBarIcon: ({ color, size }) => (
                  <MaterialCommunityIcons name="qrcode-scan" color={color} size={size} />
              ),
            }}
          />
          <Tab.Screen name="Invite" component={DeveloperScreen}
            options= {{
              tabBarLabel: 'Developers',
              tabBarIcon: ({ color, size }) => (
                  <MaterialCommunityIcons name="github-circle" color={color} size={size} />
              ),
            }}
          />
        </Tab.Navigator>
      </NavigationContainer>
    );
  }
}

async function uploadImageAsync(uri) {
  // Why are we using XMLHttpRequest? See:
  // https://github.com/expo/expo/issues/2402#issuecomment-443726662
  const blob = await new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.onload = function() {
      resolve(xhr.response);
    };
    xhr.onerror = function(e) {
      console.log(e);
      reject(new TypeError('Network request failed'));
    };
    xhr.responseType = 'blob';
    xhr.open('GET', uri, true);
    xhr.send(null);
  });

  const ref = firebase
    .storage()
    .ref()
    .child(uuid.v4());
  const snapshot = await ref.put(blob);

  // We're done with the blob, close and release it
  blob.close();

  return await snapshot.ref.getDownloadURL();
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000',
    alignItems: 'center',
    justifyContent: 'center',
  },
  postContainer: {
    flex: 1,
    backgroundColor: '#edf0f7',
    alignItems: 'center',
    justifyContent: 'center',
  },
  button: {
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    alignItems: 'center',
    shadowOpacity: 0.23,
    justifyContent: 'center',
    shadowRadius: 2.62,
    width: 150,
    height: 50,
    borderRadius: 30,
    backgroundColor:'#fff',
  },
  buttonText: {
    fontWeight: 'bold',
    color: '#fff',
  },
  setTextColor: {
    color: '#fff',
    textAlign: 'center',
    justifyContent: 'center',
  },
  stretch: {
    height: 150,
    width: 150,
    resizeMode: 'cover',
    justifyContent: "center",
    alignItems: 'center',
    alignSelf: 'center',
  },
  image: {
    flex: 1,
    resizeMode: 'cover',
    alignSelf: 'center',
    justifyContent: "center"
  },
});

export default AppScreen;
